File: scripts/plot_okpax_benchmark.m
% PLOT_OKPAX_BENCHMARK - Visualize OKPAX comprehensive benchmark results
%
% This script loads CSV files from the OKPAX benchmark and generates:
%   1. Cost convergence plots for each environment
%   2. Tree size growth analysis
%   3. Iteration timing analysis
%   4. Summary statistics comparisons
%
% USAGE:
%   1. Download the Data/Benchmarks/OKPAX/ folder from your Linux machine
%   2. Place it in the same directory as this script, or modify DATA_DIR below
%   3. Run this script in MATLAB
%
% The script will automatically find all CSV files and generate plots.

clear; close all; clc;

%% Configuration
% Modify this path to point to your downloaded benchmark data
DATA_DIR = 'Data/Benchmarks/OKPAX';

% Create output directory for plots
OUTPUT_DIR = fullfile(DATA_DIR, 'plots');
if ~exist(OUTPUT_DIR, 'dir')
    mkdir(OUTPUT_DIR);
end

fprintf('OKPAX Benchmark Visualization\n');
fprintf('==============================\n\n');

%% Load Data
fprintf('Loading benchmark data from: %s\n', DATA_DIR);

% Find summary file
summary_files = dir(fullfile(DATA_DIR, '*_summary.csv'));
if isempty(summary_files)
    error('No summary CSV file found in %s', DATA_DIR);
end

% Load summary data
summary_file = fullfile(DATA_DIR, summary_files(1).name);
summary_data = readtable(summary_file);
fprintf('✓ Loaded summary: %d runs\n', height(summary_data));

% Find detailed iteration files
detail_files = dir(fullfile(DATA_DIR, '*_run*.csv'));
fprintf('✓ Found %d detailed iteration files\n', length(detail_files));

% Load all detailed data
all_detailed_data = [];
for i = 1:length(detail_files)
    filename = detail_files(i).name;
    filepath = fullfile(DATA_DIR, filename);

    % Parse metadata from filename
    % Format: {timestamp}_{environment}_{planner_type}_run{N}.csv
    parts = split(filename, '_');

    % Extract environment, planner type, and run number
    % Assumes format: okpax_benchmark_{timestamp}_{environment}_{planner}_run{N}.csv
    if length(parts) >= 4
        environment = parts{end-2};
        planner_type = parts{end-1};
        run_str = parts{end};
        run_number = str2double(replace(run_str, {'.csv', 'run'}, ''));

        % Load CSV
        data = readtable(filepath);

        % Add metadata columns
        data.environment = repmat({environment}, height(data), 1);
        data.planner_type = repmat({planner_type}, height(data), 1);
        data.run_number = repmat(run_number, height(data), 1);

        all_detailed_data = [all_detailed_data; data];
    end
end

fprintf('✓ Loaded %d total iteration records\n\n', height(all_detailed_data));

% Get unique values
environments = unique(all_detailed_data.environment);
planner_types = unique(all_detailed_data.planner_type);

fprintf('Environments: %s\n', strjoin(environments, ', '));
fprintf('Planner types: %s\n\n', strjoin(planner_types, ', '));

%% Generate Plots for Each Environment
fprintf('Generating plots...\n\n');

for env_idx = 1:length(environments)
    env = environments{env_idx};
    fprintf('Processing environment: %s\n', env);

    % Filter data for this environment
    env_mask = strcmp(all_detailed_data.environment, env);
    env_data = all_detailed_data(env_mask, :);

    % Create figure with subplots
    fig = figure('Position', [100, 100, 1600, 1200]);
    sgtitle(sprintf('OKPAX Performance Analysis - %s Environment', env), ...
            'FontSize', 16, 'FontWeight', 'bold');

    %% Subplot 1: Cost Convergence (All Runs + Mean)
    subplot(2, 2, 1);
    hold on; grid on;

    colors = lines(length(planner_types));

    for p_idx = 1:length(planner_types)
        planner = planner_types{p_idx};
        planner_mask = strcmp(env_data.planner_type, planner);
        planner_data = env_data(planner_mask, :);

        % Plot individual runs (transparent)
        runs = unique(planner_data.run_number);
        for r_idx = 1:length(runs)
            run = runs(r_idx);
            run_mask = planner_data.run_number == run;
            run_data = planner_data(run_mask, :);

            plot(run_data.iteration, run_data.min_cost, ...
                 'Color', [colors(p_idx, :), 0.3], 'LineWidth', 1);
        end

        % Plot mean (bold)
        [G, iterations] = findgroups(planner_data.iteration);
        mean_cost = splitapply(@mean, planner_data.min_cost, G);

        plot(iterations, mean_cost, 'Color', colors(p_idx, :), ...
             'LineWidth', 3, 'DisplayName', planner);
    end

    xlabel('Iteration');
    ylabel('Minimum Cost');
    title('Cost Convergence (all runs + mean)');
    legend('Location', 'best');
    hold off;

    %% Subplot 2: Mean Cost with Confidence Intervals
    subplot(2, 2, 2);
    hold on; grid on;

    for p_idx = 1:length(planner_types)
        planner = planner_types{p_idx};
        planner_mask = strcmp(env_data.planner_type, planner);
        planner_data = env_data(planner_mask, :);

        % Compute mean and std by iteration
        [G, iterations] = findgroups(planner_data.iteration);
        mean_cost = splitapply(@mean, planner_data.min_cost, G);
        std_cost = splitapply(@std, planner_data.min_cost, G);

        % Plot mean line
        plot(iterations, mean_cost, 'Color', colors(p_idx, :), ...
             'LineWidth', 2, 'DisplayName', planner);

        % Plot confidence interval
        fill([iterations; flipud(iterations)], ...
             [mean_cost - std_cost; flipud(mean_cost + std_cost)], ...
             colors(p_idx, :), 'FaceAlpha', 0.2, 'EdgeColor', 'none', ...
             'HandleVisibility', 'off');
    end

    xlabel('Iteration');
    ylabel('Minimum Cost');
    title('Mean Cost ± Std Dev');
    legend('Location', 'best');
    hold off;

    %% Subplot 3: Tree Size Growth
    subplot(2, 2, 3);
    hold on; grid on;

    for p_idx = 1:length(planner_types)
        planner = planner_types{p_idx};
        planner_mask = strcmp(env_data.planner_type, planner);
        planner_data = env_data(planner_mask, :);

        % Compute mean tree size by iteration
        [G, iterations] = findgroups(planner_data.iteration);
        mean_tree_size = splitapply(@mean, planner_data.tree_size, G);

        plot(iterations, mean_tree_size, 'Color', colors(p_idx, :), ...
             'LineWidth', 2, 'DisplayName', planner);
    end

    xlabel('Iteration');
    ylabel('Tree Size');
    title('Mean Tree Size Growth');
    legend('Location', 'best');
    hold off;

    %% Subplot 4: Iteration Execution Time
    subplot(2, 2, 4);
    hold on; grid on;

    for p_idx = 1:length(planner_types)
        planner = planner_types{p_idx};
        planner_mask = strcmp(env_data.planner_type, planner);
        planner_data = env_data(planner_mask, :);

        % Compute mean execution time by iteration
        [G, iterations] = findgroups(planner_data.iteration);
        mean_time = splitapply(@mean, planner_data.execution_time_ms, G);

        plot(iterations, mean_time, 'Color', colors(p_idx, :), ...
             'LineWidth', 2, 'DisplayName', planner);
    end

    xlabel('Iteration');
    ylabel('Time (ms)');
    title('Mean Iteration Execution Time');
    legend('Location', 'best');
    hold off;

    % Save figure
    output_file = fullfile(OUTPUT_DIR, sprintf('cost_convergence_%s.png', env));
    saveas(fig, output_file);
    fprintf('  ✓ Saved: %s\n', output_file);

    % Also save as .fig for interactive viewing
    saveas(fig, fullfile(OUTPUT_DIR, sprintf('cost_convergence_%s.fig', env)));
end

%% Summary Statistics Plots
fprintf('\nGenerating summary statistics plots...\n');

fig = figure('Position', [100, 100, 1600, 1200]);
sgtitle('OKPAX Summary Statistics', 'FontSize', 16, 'FontWeight', 'bold');

% Create grouping variable for box plots
summary_data.group = strcat(summary_data.environment, '_', summary_data.planner_type);

%% Subplot 1: Final Cost Distribution
subplot(2, 2, 1);
boxplot(summary_data.final_cost, summary_data.group);
ylabel('Final Cost');
title('Final Cost Distribution');
xtickangle(45);
grid on;

%% Subplot 2: Total Time Comparison
subplot(2, 2, 2);
boxplot(summary_data.total_time_seconds, summary_data.group);
ylabel('Time (seconds)');
title('Total Execution Time');
xtickangle(45);
grid on;

%% Subplot 3: Final Iterations
subplot(2, 2, 3);
boxplot(summary_data.final_iterations, summary_data.group);
ylabel('Iterations');
title('Iterations to Convergence');
xtickangle(45);
grid on;

%% Subplot 4: Final Tree Size
subplot(2, 2, 4);
boxplot(summary_data.final_tree_size, summary_data.group);
ylabel('Tree Size');
title('Final Tree Size');
xtickangle(45);
grid on;

% Save summary figure
output_file = fullfile(OUTPUT_DIR, 'summary_statistics.png');
saveas(fig, output_file);
fprintf('✓ Saved: %s\n', output_file);
saveas(fig, fullfile(OUTPUT_DIR, 'summary_statistics.fig'));

%% Generate Text Report
fprintf('\nGenerating text report...\n');

report_file = fullfile(OUTPUT_DIR, 'benchmark_report.txt');
fid = fopen(report_file, 'w');

fprintf(fid, 'OKPAX Comprehensive Benchmark Report\n');
fprintf(fid, '=====================================\n\n');

for env_idx = 1:length(environments)
    env = environments{env_idx};
    fprintf(fid, 'Environment: %s\n', env);
    fprintf(fid, '%s\n\n', repmat('-', 1, 50));

    env_mask = strcmp(summary_data.environment, env);
    env_summary = summary_data(env_mask, :);

    for p_idx = 1:length(planner_types)
        planner = planner_types{p_idx};
        planner_mask = strcmp(env_summary.planner_type, planner);
        planner_summary = env_summary(planner_mask, :);

        if isempty(planner_summary)
            continue;
        end

        fprintf(fid, '  %s:\n', planner);
        fprintf(fid, '    Runs: %d\n', height(planner_summary));
        fprintf(fid, '    Mean Final Cost: %.3f ± %.3f\n', ...
                mean(planner_summary.final_cost), std(planner_summary.final_cost));
        fprintf(fid, '    Mean Total Time: %.3fs ± %.3fs\n', ...
                mean(planner_summary.total_time_seconds), std(planner_summary.total_time_seconds));
        fprintf(fid, '    Mean Iterations: %.1f ± %.1f\n', ...
                mean(planner_summary.final_iterations), std(planner_summary.final_iterations));
        fprintf(fid, '    Mean Final Tree Size: %.0f ± %.0f\n', ...
                mean(planner_summary.final_tree_size), std(planner_summary.final_tree_size));
        fprintf(fid, '    Best Cost: %.3f\n', min(planner_summary.final_cost));
        fprintf(fid, '    Worst Cost: %.3f\n\n', max(planner_summary.final_cost));
    end

    fprintf(fid, '\n');
end

fclose(fid);
fprintf('✓ Saved: %s\n', report_file);

%% Summary Table
fprintf('\n');
fprintf('Summary Statistics by Environment and Planner\n');
fprintf('==============================================\n\n');

for env_idx = 1:length(environments)
    env = environments{env_idx};
    fprintf('Environment: %s\n', env);

    env_mask = strcmp(summary_data.environment, env);
    env_summary = summary_data(env_mask, :);

    % Create summary table
    T = table();
    for p_idx = 1:length(planner_types)
        planner = planner_types{p_idx};
        planner_mask = strcmp(env_summary.planner_type, planner);
        planner_summary = env_summary(planner_mask, :);

        if isempty(planner_summary)
            continue;
        end

        row = table();
        row.Planner = {planner};
        row.Runs = height(planner_summary);
        row.MeanCost = mean(planner_summary.final_cost);
        row.StdCost = std(planner_summary.final_cost);
        row.MeanTime = mean(planner_summary.total_time_seconds);
        row.StdTime = std(planner_summary.total_time_seconds);
        row.MeanIters = mean(planner_summary.final_iterations);
        row.MeanTreeSize = mean(planner_summary.final_tree_size);

        T = [T; row];
    end

    disp(T);
    fprintf('\n');
end

%% Final Summary
fprintf('═══════════════════════════════════════════\n');
fprintf('All plots saved to: %s\n', OUTPUT_DIR);
fprintf('═══════════════════════════════════════════\n');
fprintf('\nGenerated files:\n');
fprintf('  - cost_convergence_{environment}.png/.fig\n');
fprintf('  - summary_statistics.png/.fig\n');
fprintf('  - benchmark_report.txt\n');
fprintf('\nDone!\n');

File: scripts/plot_okpax_benchmark.py
#!/usr/bin/env python3
"""
Visualization script for OKPAX comprehensive benchmark results.
Plots cost improvements over iterations and compares planner performance.
"""

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import glob
import os
import sys
from pathlib import Path
import numpy as np

# Set style
sns.set_style("whitegrid")
sns.set_palette("husl")
plt.rcParams['figure.figsize'] = (14, 10)
plt.rcParams['font.size'] = 10

def load_benchmark_data(benchmark_dir):
    """Load all benchmark CSV files from a directory."""

    # Load summary
    summary_files = glob.glob(os.path.join(benchmark_dir, "*_summary.csv"))
    if not summary_files:
        print(f"No summary file found in {benchmark_dir}")
        return None, []

    summary_df = pd.read_csv(summary_files[0])

    # Load detailed iteration files
    detail_files = glob.glob(os.path.join(benchmark_dir, "*_run*.csv"))

    detailed_data = []
    for file in detail_files:
        filename = os.path.basename(file)
        parts = filename.replace('.csv', '').split('_')

        # Extract metadata from filename
        # Format: {timestamp}_{environment}_{planner_type}_run{N}.csv
        if len(parts) >= 4:
            environment = parts[-3]
            planner_type = parts[-2]
            run_number = int(parts[-1].replace('run', ''))

            df = pd.read_csv(file)
            df['environment'] = environment
            df['planner_type'] = planner_type
            df['run_number'] = run_number

            detailed_data.append(df)

    return summary_df, detailed_data

def plot_cost_convergence(detailed_data, output_dir):
    """Plot cost convergence over iterations for each environment."""

    if not detailed_data:
        print("No detailed data to plot")
        return

    # Combine all data
    all_data = pd.concat(detailed_data, ignore_index=True)

    environments = all_data['environment'].unique()
    planner_types = all_data['planner_type'].unique()

    for env in environments:
        fig, axes = plt.subplots(2, 2, figsize=(16, 12))
        fig.suptitle(f'OKPAX Performance Analysis - {env} Environment', fontsize=16, fontweight='bold')

        env_data = all_data[all_data['environment'] == env]

        # Plot 1: Cost over iterations (all runs)
        ax = axes[0, 0]
        for planner in planner_types:
            planner_data = env_data[env_data['planner_type'] == planner]
            for run in planner_data['run_number'].unique():
                run_data = planner_data[planner_data['run_number'] == run]
                ax.plot(run_data['iteration'], run_data['min_cost'],
                       alpha=0.3, linewidth=1)

        # Plot mean
        for planner in planner_types:
            planner_data = env_data[env_data['planner_type'] == planner]
            mean_data = planner_data.groupby('iteration')['min_cost'].mean().reset_index()
            ax.plot(mean_data['iteration'], mean_data['min_cost'],
                   label=planner, linewidth=3, alpha=0.9)

        ax.set_xlabel('Iteration')
        ax.set_ylabel('Minimum Cost')
        ax.set_title('Cost Convergence (all runs + mean)')
        ax.legend()
        ax.grid(True, alpha=0.3)

        # Plot 2: Mean cost with confidence intervals
        ax = axes[0, 1]
        for planner in planner_types:
            planner_data = env_data[env_data['planner_type'] == planner]
            grouped = planner_data.groupby('iteration')['min_cost']
            mean = grouped.mean()
            std = grouped.std()

            ax.plot(mean.index, mean.values, label=planner, linewidth=2)
            ax.fill_between(mean.index,
                           mean.values - std.values,
                           mean.values + std.values,
                           alpha=0.2)

        ax.set_xlabel('Iteration')
        ax.set_ylabel('Minimum Cost')
        ax.set_title('Mean Cost ± Std Dev')
        ax.legend()
        ax.grid(True, alpha=0.3)

        # Plot 3: Tree size growth
        ax = axes[1, 0]
        for planner in planner_types:
            planner_data = env_data[env_data['planner_type'] == planner]
            mean_data = planner_data.groupby('iteration')['tree_size'].mean().reset_index()
            ax.plot(mean_data['iteration'], mean_data['tree_size'],
                   label=planner, linewidth=2)

        ax.set_xlabel('Iteration')
        ax.set_ylabel('Tree Size')
        ax.set_title('Mean Tree Size Growth')
        ax.legend()
        ax.grid(True, alpha=0.3)

        # Plot 4: Iteration execution time
        ax = axes[1, 1]
        for planner in planner_types:
            planner_data = env_data[env_data['planner_type'] == planner]
            mean_data = planner_data.groupby('iteration')['execution_time_ms'].mean().reset_index()
            ax.plot(mean_data['iteration'], mean_data['execution_time_ms'],
                   label=planner, linewidth=2)

        ax.set_xlabel('Iteration')
        ax.set_ylabel('Time (ms)')
        ax.set_title('Mean Iteration Execution Time')
        ax.legend()
        ax.grid(True, alpha=0.3)

        plt.tight_layout()
        output_file = os.path.join(output_dir, f'cost_convergence_{env}.png')
        plt.savefig(output_file, dpi=300, bbox_inches='tight')
        print(f"Saved: {output_file}")
        plt.close()

def plot_summary_statistics(summary_df, output_dir):
    """Plot summary statistics comparing planners."""

    if summary_df is None or summary_df.empty:
        print("No summary data to plot")
        return

    fig, axes = plt.subplots(2, 2, figsize=(16, 12))
    fig.suptitle('OKPAX Summary Statistics', fontsize=16, fontweight='bold')

    # Plot 1: Final cost distribution
    ax = axes[0, 0]
    summary_df.boxplot(column='final_cost', by=['environment', 'planner_type'], ax=ax)
    ax.set_title('Final Cost Distribution')
    ax.set_ylabel('Final Cost')
    plt.sca(ax)
    plt.xticks(rotation=45, ha='right')

    # Plot 2: Total time comparison
    ax = axes[0, 1]
    summary_df.boxplot(column='total_time_seconds', by=['environment', 'planner_type'], ax=ax)
    ax.set_title('Total Execution Time')
    ax.set_ylabel('Time (seconds)')
    plt.sca(ax)
    plt.xticks(rotation=45, ha='right')

    # Plot 3: Final iterations
    ax = axes[1, 0]
    summary_df.boxplot(column='final_iterations', by=['environment', 'planner_type'], ax=ax)
    ax.set_title('Iterations to Convergence')
    ax.set_ylabel('Iterations')
    plt.sca(ax)
    plt.xticks(rotation=45, ha='right')

    # Plot 4: Final tree size
    ax = axes[1, 1]
    summary_df.boxplot(column='final_tree_size', by=['environment', 'planner_type'], ax=ax)
    ax.set_title('Final Tree Size')
    ax.set_ylabel('Tree Size')
    plt.sca(ax)
    plt.xticks(rotation=45, ha='right')

    plt.tight_layout()
    output_file = os.path.join(output_dir, 'summary_statistics.png')
    plt.savefig(output_file, dpi=300, bbox_inches='tight')
    print(f"Saved: {output_file}")
    plt.close()

def generate_markdown_report(summary_df, detailed_data, output_dir):
    """Generate a markdown report with statistics."""

    if summary_df is None:
        return

    report_file = os.path.join(output_dir, 'benchmark_report.md')

    with open(report_file, 'w') as f:
        f.write("# OKPAX Comprehensive Benchmark Report\n\n")
        f.write("## Summary Statistics\n\n")

        for env in summary_df['environment'].unique():
            f.write(f"### Environment: {env}\n\n")
            env_data = summary_df[summary_df['environment'] == env]

            for planner in env_data['planner_type'].unique():
                planner_data = env_data[env_data['planner_type'] == planner]

                f.write(f"#### {planner}\n\n")
                f.write(f"- **Runs**: {len(planner_data)}\n")
                f.write(f"- **Mean Final Cost**: {planner_data['final_cost'].mean():.3f} ± {planner_data['final_cost'].std():.3f}\n")
                f.write(f"- **Mean Total Time**: {planner_data['total_time_seconds'].mean():.3f}s ± {planner_data['total_time_seconds'].std():.3f}s\n")
                f.write(f"- **Mean Iterations**: {planner_data['final_iterations'].mean():.1f} ± {planner_data['final_iterations'].std():.1f}\n")
                f.write(f"- **Mean Final Tree Size**: {planner_data['final_tree_size'].mean():.0f} ± {planner_data['final_tree_size'].std():.0f}\n")
                f.write(f"- **Best Cost**: {planner_data['final_cost'].min():.3f}\n")
                f.write(f"- **Worst Cost**: {planner_data['final_cost'].max():.3f}\n\n")

    print(f"Saved: {report_file}")

def main():
    # Find the most recent benchmark directory
    benchmark_base = Path("Data/Benchmarks/OKPAX")

    if not benchmark_base.exists():
        print(f"Error: Benchmark directory not found: {benchmark_base}")
        print("Please run the benchmark first: ./OKPAXComprehensiveBenchmark")
        sys.exit(1)

    # Create output directory for plots
    output_dir = benchmark_base / "plots"
    output_dir.mkdir(exist_ok=True)

    print(f"Loading benchmark data from {benchmark_base}...")
    summary_df, detailed_data = load_benchmark_data(str(benchmark_base))

    if summary_df is None:
        print("Error: No benchmark data found")
        sys.exit(1)

    print(f"Loaded {len(summary_df)} benchmark runs")
    print(f"Loaded {len(detailed_data)} detailed iteration files")

    # Generate plots
    print("\nGenerating convergence plots...")
    plot_cost_convergence(detailed_data, str(output_dir))

    print("\nGenerating summary statistics plots...")
    plot_summary_statistics(summary_df, str(output_dir))

    print("\nGenerating markdown report...")
    generate_markdown_report(summary_df, detailed_data, str(output_dir))

    print(f"\n✓ All plots saved to: {output_dir}")
    print("\nDone!")

if __name__ == "__main__":
    main()
